# Global Weather Data Analysis
- This is an analysis project to understand the weather of countries in 2024

In [350]:
import pandas as pd               # data handling
import matplotlib.pyplot as plt   # visualization
import seaborn as sns             # visualization

- Pandas for data manipulation
- matplotlib and seaborn for visualization

## Loading Data

In [351]:
df= pd.read_csv('weather.csv')
df

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1715849100,2024-05-16 13:15,26.6,79.8,Partly Cloudy,...,8.40,26.60,1,1,4:50 AM,6:50 PM,12:12 PM,1:11 AM,Waxing Gibbous,55
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1715849100,2024-05-16 10:45,19.0,66.2,Partly cloudy,...,1.10,2.00,1,1,5:21 AM,7:54 PM,12:58 PM,2:14 AM,Waxing Gibbous,55
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 9:45,23.0,73.4,Sunny,...,10.40,18.40,1,1,5:40 AM,7:50 PM,1:15 PM,2:14 AM,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.70,0.90,1,1,6:31 AM,9:11 PM,2:12 PM,3:31 AM,Waxing Gibbous,55
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1715849100,2024-05-16 9:45,26.0,78.8,Partly cloudy,...,183.40,262.30,5,10,6:12 AM,5:55 PM,1:17 PM,12:38 AM,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 2:30,17.5,63.6,Partly Cloudy,...,7.85,9.35,1,1,6:14 AM,6:39 PM,8:54 AM,10:12 PM,Waxing Crescent,10
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30,28.3,82.9,Cloudy,...,36.25,39.45,2,4,5:34 AM,6:17 PM,7:30 AM,9:34 PM,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 9:30,21.0,69.7,Sunny,...,12.85,51.65,1,2,5:46 AM,6:18 PM,8:00 AM,9:39 PM,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 8:30,21.5,70.8,Sunny,...,5.55,5.85,1,1,6:15 AM,5:56 PM,9:13 AM,8:35 PM,Waxing Crescent,9


## Inspecting Data

In [352]:
df.head()

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.52,69.18,Asia/Kabul,1715849100,2024-05-16 13:15,26.6,79.8,Partly Cloudy,...,8.4,26.6,1,1,4:50 AM,6:50 PM,12:12 PM,1:11 AM,Waxing Gibbous,55
1,Albania,Tirana,41.33,19.82,Europe/Tirane,1715849100,2024-05-16 10:45,19.0,66.2,Partly cloudy,...,1.1,2.0,1,1,5:21 AM,7:54 PM,12:58 PM,2:14 AM,Waxing Gibbous,55
2,Algeria,Algiers,36.76,3.05,Africa/Algiers,1715849100,2024-05-16 9:45,23.0,73.4,Sunny,...,10.4,18.4,1,1,5:40 AM,7:50 PM,1:15 PM,2:14 AM,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.50,1.52,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.7,0.9,1,1,6:31 AM,9:11 PM,2:12 PM,3:31 AM,Waxing Gibbous,55
4,Angola,Luanda,-8.84,13.23,Africa/Luanda,1715849100,2024-05-16 9:45,26.0,78.8,Partly cloudy,...,183.4,262.3,5,10,6:12 AM,5:55 PM,1:17 PM,12:38 AM,Waxing Gibbous,55


In [353]:
df.tail(10)

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
136623,United Kingdom,London,51.5171,-0.1062,Europe/London,1776666600,2026-04-20 7:30,3.2,37.8,Sunny,...,14.35,18.85,1,2,5:54 AM,8:05 PM,6:54 AM,No moonset,Waxing Crescent,9
136624,United States of America,Washington Harbor,48.0792,-123.0439,America/Los_Angeles,1776666600,2026-04-19 23:30,13.5,56.3,Clear,...,5.45,9.65,1,1,6:13 AM,8:10 PM,7:44 AM,12:01 AM,Waxing Crescent,11
136625,Uruguay,Montevideo,-34.8581,-56.1708,America/Montevideo,1776666600,2026-04-20 3:30,20.3,68.5,Overcast,...,10.25,12.35,1,1,7:12 AM,6:14 PM,11:01 AM,8:32 PM,Waxing Crescent,10
136626,Uzbekistan,Tashkent,41.3167,69.2500,Asia/Tashkent,1776666600,2026-04-20 11:30,22.9,73.3,Partly Cloudy,...,11.45,13.65,1,1,5:37 AM,7:08 PM,7:01 AM,11:14 PM,Waxing Crescent,8
136627,Vanuatu,Port Vila,-17.7333,168.3167,Pacific/Efate,1776666600,2026-04-20 17:30,26.2,79.2,Overcast,...,6.45,9.65,1,1,5:57 AM,5:35 PM,8:31 AM,7:46 PM,Waxing Crescent,6
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 2:30,17.5,63.6,Partly Cloudy,...,7.85,9.35,1,1,6:14 AM,6:39 PM,8:54 AM,10:12 PM,Waxing Crescent,10
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30,28.3,82.9,Cloudy,...,36.25,39.45,2,4,5:34 AM,6:17 PM,7:30 AM,9:34 PM,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 9:30,21.0,69.7,Sunny,...,12.85,51.65,1,2,5:46 AM,6:18 PM,8:00 AM,9:39 PM,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 8:30,21.5,70.8,Sunny,...,5.55,5.85,1,1,6:15 AM,5:56 PM,9:13 AM,8:35 PM,Waxing Crescent,9
136632,Zimbabwe,Harare,-17.8178,31.0447,Africa/Harare,1776666600,2026-04-20 8:30,20.0,68.1,Sunny,...,7.85,8.05,1,1,6:06 AM,5:43 PM,9:07 AM,8:18 PM,Waxing Crescent,9


In [354]:
df.sample(3)

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
80103,Belize,Belmopan,17.2500,-88.7667,America/Belize,1751532300,2025-07-03 2:45,23.4,74.1,Fog,...,29.23,61.42,2,3,5:26 AM,6:33 PM,12:41 PM,No moonset,Waxing Gibbous,54
119240,Thailand,Bangkok,13.7500,100.5167,Asia/Bangkok,1768893300,2026-01-20 14:15,32.4,90.3,Sunny,...,26.55,27.65,2,3,6:46 AM,6:12 PM,7:48 AM,7:34 PM,Waxing Crescent,1
133178,Estonia,Tallinn,59.4339,24.7281,Europe/Tallinn,1775197800,2026-04-03 9:30,4.2,39.6,Light rain,...,6.95,8.95,1,1,6:42 AM,8:08 PM,10:37 PM,6:28 AM,Waning Gibbous,99


### Checking Data Types

In [355]:
df.dtypes
# check column data type

country                          object
location_name                    object
latitude                        float64
longitude                       float64
timezone                         object
last_updated_epoch                int64
last_updated                     object
temperature_celsius             float64
temperature_fahrenheit          float64
condition_text                   object
wind_mph                        float64
wind_kph                        float64
wind_degree                       int64
wind_direction                   object
pressure_mb                       int64
pressure_in                     float64
precip_mm                       float64
precip_in                       float64
humidity                          int64
cloud                             int64
feels_like_celsius              float64
feels_like_fahrenheit           float64
visibility_km                   float64
visibility_miles                  int64
uv_index                        float64


In [356]:
df.columns
# Display Column Names

Index(['country', 'location_name', 'latitude', 'longitude', 'timezone',
       'last_updated_epoch', 'last_updated', 'temperature_celsius',
       'temperature_fahrenheit', 'condition_text', 'wind_mph', 'wind_kph',
       'wind_degree', 'wind_direction', 'pressure_mb', 'pressure_in',
       'precip_mm', 'precip_in', 'humidity', 'cloud', 'feels_like_celsius',
       'feels_like_fahrenheit', 'visibility_km', 'visibility_miles',
       'uv_index', 'gust_mph', 'gust_kph', 'air_quality_Carbon_Monoxide',
       'air_quality_Ozone', 'air_quality_Nitrogen_dioxide',
       'air_quality_Sulphur_dioxide', 'air_quality_PM2.5', 'air_quality_PM10',
       'air_quality_us-epa-index', 'air_quality_gb-defra-index', 'sunrise',
       'sunset', 'moonrise', 'moonset', 'moon_phase', 'moon_illumination'],
      dtype='object')

In [357]:
df.shape
# Display number or rows and columns

(136633, 41)

In [358]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 136633 entries, 0 to 136632
Data columns (total 41 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   country                       136633 non-null  object 
 1   location_name                 136633 non-null  object 
 2   latitude                      136633 non-null  float64
 3   longitude                     136633 non-null  float64
 4   timezone                      136633 non-null  object 
 5   last_updated_epoch            136633 non-null  int64  
 6   last_updated                  136633 non-null  object 
 7   temperature_celsius           136633 non-null  float64
 8   temperature_fahrenheit        136633 non-null  float64
 9   condition_text                136633 non-null  object 
 10  wind_mph                      136633 non-null  float64
 11  wind_kph                      136633 non-null  float64
 12  wind_degree                   136633 non-nul

### Dataset Overview
- number of rows = 136633
- number of columns = 41
- Data Types: float = 21, integer = 9, objects = 11
- missing values = none
- sunrise, moonset, moonrise and sunset times stored as object
- last_updated stored as object
- memory used = 42.7+ MB

### Selecting Columns

- selecting single columns

In [359]:
df['country']

0         Afghanistan
1             Albania
2             Algeria
3             Andorra
4              Angola
             ...     
136628      Venezuela
136629        Vietnam
136630          Yemen
136631         Zambia
136632       Zimbabwe
Name: country, Length: 136633, dtype: object

In [360]:
df['sunrise']

0         4:50 AM
1         5:21 AM
2         5:40 AM
3         6:31 AM
4         6:12 AM
           ...   
136628    6:14 AM
136629    5:34 AM
136630    5:46 AM
136631    6:15 AM
136632    6:06 AM
Name: sunrise, Length: 136633, dtype: object

- selecting multiple columns

In [361]:
df[['sunrise','sunset']]

,sunrise,sunset
0,4:50 AM,6:50 PM
1,5:21 AM,7:54 PM
2,5:40 AM,7:50 PM
3,6:31 AM,9:11 PM
4,6:12 AM,5:55 PM
...,...,...
136628,6:14 AM,6:39 PM
136629,5:34 AM,6:17 PM
136630,5:46 AM,6:18 PM
136631,6:15 AM,5:56 PM


In [362]:
place = df[['country','location_name']]

In [363]:
place

,country,location_name
0,Afghanistan,Kabul
1,Albania,Tirana
2,Algeria,Algiers
3,Andorra,Andorra La Vella
4,Angola,Luanda
...,...,...
136628,Venezuela,Caracas
136629,Vietnam,Hanoi
136630,Yemen,Sanaa
136631,Zambia,Lusaka


### Filtering the  data

### 1

In [364]:
df[df['temperature_celsius']==25]

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
31,Cameroon,Yaounde,3.8700,11.5200,Africa/Douala,1715849100,2024-05-16 9:45,25.0,76.9,Partly Cloudy,...,25.60,31.20,2,3,6:01 AM,6:19 PM,1:13 PM,12:59 AM,Waxing Gibbous,55
43,Cyprus,Nicosia,35.1700,33.3700,Asia/Famagusta,1715849100,2024-05-16 11:45,25.0,77.0,Partly cloudy,...,7.70,13.50,1,1,5:42 AM,7:45 PM,1:11 PM,2:08 AM,Waxing Gibbous,55
82,Israel,Jerusalem,31.7800,35.2300,Asia/Jerusalem,1715849100,2024-05-16 11:45,25.0,77.0,Partly cloudy,...,5.00,9.40,1,1,5:42 AM,7:30 PM,1:07 PM,1:55 AM,Waxing Gibbous,55
144,Rwanda,Kigali,-1.9500,30.0600,Africa/Kigali,1715849100,2024-05-16 10:45,25.0,77.0,Partly cloudy,...,4.20,6.30,1,1,5:55 AM,5:57 PM,1:02 PM,12:36 AM,Waxing Gibbous,55
172,Tanzania,Dodoma,-6.1800,35.7500,Africa/Dar_es_Salaam,1715849100,2024-05-16 11:45,25.0,77.0,Partly cloudy,...,1.90,3.40,1,1,6:38 AM,6:28 PM,1:43 PM,1:08 AM,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136101,Equatorial Guinea,Malabo,3.7500,8.7833,Africa/Malabo,1776493800,2026-04-18 7:30,25.0,77.0,Patchy light rain with thunder,...,6.65,9.75,1,1,6:18 AM,6:30 PM,6:54 AM,7:33 PM,Waxing Crescent,0
136389,Saint Lucia,Castries,14.0000,-61.0000,America/St_Lucia,1776579300,2026-04-19 2:15,25.0,77.0,Partly cloudy,...,8.55,13.35,1,1,5:48 AM,6:18 PM,7:22 AM,8:48 PM,Waxing Crescent,4
136390,Saint Vincent and the Grenadines,Kingstown,13.1333,-61.2167,America/St_Vincent,1776579300,2026-04-19 2:15,25.0,77.0,Partly cloudy,...,6.60,10.00,1,1,5:50 AM,6:18 PM,7:25 AM,8:47 PM,Waxing Crescent,4
136483,Democratic Republic of Congo,Kinshasa,-4.3000,15.3000,Africa/Kinshasa,1776666600,2026-04-20 7:30,25.0,77.1,Patchy rain nearby,...,14.75,14.95,1,2,5:58 AM,5:58 PM,8:42 AM,8:55 PM,Waxing Crescent,9


### 2

In [365]:
df[df['moon_phase']=='New Moon']

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
4285,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1717682400,2024-06-06 18:30,18.0,64.3,Patchy rain nearby,...,3.90,8.20,1,1,4:40 AM,7:04 PM,4:02 AM,7:18 PM,New Moon,0
4286,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1717682400,2024-06-06 16:00,32.0,89.6,Sunny,...,4.50,5.10,1,1,5:08 AM,8:11 PM,4:31 AM,8:41 PM,New Moon,0
4287,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1717682400,2024-06-06 15:00,32.0,89.6,Sunny,...,16.40,34.80,2,2,5:29 AM,8:04 PM,4:58 AM,8:33 PM,New Moon,0
4288,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1717682400,2024-06-06 16:00,22.5,72.6,Patchy rain nearby,...,2.20,3.40,1,1,6:17 AM,9:29 PM,5:43 AM,10:03 PM,New Moon,0
4289,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1717682400,2024-06-06 15:00,26.0,78.8,Partly cloudy,...,7.30,25.20,1,1,6:17 AM,5:55 PM,6:04 AM,5:55 PM,New Moon,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136196,Samoa,Apia,-13.8333,-171.7333,Pacific/Apia,1776493800,2026-04-18 19:30,27.4,81.3,Partly cloudy,...,4.65,6.75,1,1,6:33 AM,6:19 PM,6:50 AM,6:40 PM,New Moon,0
136207,Solomon Islands,Honiara,-9.4333,159.9500,Pacific/Guadalcanal,1776493800,2026-04-18 17:30,31.2,88.2,Patchy rain nearby,...,2.45,2.95,1,1,6:23 AM,6:16 PM,6:42 AM,6:44 PM,New Moon,0
136224,Tonga,Nuku`Aloia,-21.1333,-175.2000,Pacific/Tongatapu,1776493800,2026-04-18 19:30,24.1,75.4,Clear,...,6.15,9.45,1,1,6:53 AM,6:27 PM,7:14 AM,6:43 PM,New Moon,0
136229,Tuvalu,Funafuti,-8.5167,179.2167,Pacific/Funafuti,1776493800,2026-04-18 18:30,30.3,86.5,Partly cloudy,...,1.75,2.25,1,1,6:06 AM,5:59 PM,6:21 AM,6:25 PM,New Moon,0


### 3

In [366]:
df[df['condition_text']=='Patchy light rain with thunder']

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
24,Brunei Darussalam,Bandar Seri Begawan,4.8800,114.9300,Asia/Brunei,1715849100,2024-05-16 16:45,32.0,89.6,Patchy light rain with thunder,...,2.20,2.30,1,1,6:06 AM,6:27 PM,1:06 PM,12:54 AM,Waxing Gibbous,55
97,Thailand,Nan,18.7800,100.7800,Asia/Bangkok,1715849100,2024-05-16 15:45,33.0,91.4,Patchy light rain with thunder,...,2.80,3.00,1,1,5:43 AM,6:44 PM,12:50 PM,1:10 AM,Waxing Gibbous,55
219,Brunei Darussalam,Bandar Seri Begawan,4.8800,114.9300,Asia/Brunei,1715868000,2024-05-16 22:00,27.0,80.6,Patchy light rain with thunder,...,9.50,9.80,1,1,6:06 AM,6:27 PM,1:06 PM,12:54 AM,Waxing Gibbous,55
234,Congo,Brazzaville,-4.2600,15.2800,Africa/Brazzaville,1715868000,2024-05-16 15:00,32.0,89.6,Patchy light rain with thunder,...,15.00,31.10,1,2,5:58 AM,5:53 PM,1:05 PM,12:34 AM,Waxing Gibbous,55
240,Democratic Republic of Congo,Kinshasa,-4.3000,15.3000,Africa/Brazzaville,1715868000,2024-05-16 15:00,32.0,89.6,Patchy light rain with thunder,...,15.00,31.10,1,2,5:58 AM,5:53 PM,1:05 PM,12:34 AM,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136101,Equatorial Guinea,Malabo,3.7500,8.7833,Africa/Malabo,1776493800,2026-04-18 7:30,25.0,77.0,Patchy light rain with thunder,...,6.65,9.75,1,1,6:18 AM,6:30 PM,6:54 AM,7:33 PM,Waxing Crescent,0
136274,Cameroon,Douala,4.0503,9.7000,Africa/Douala,1776579300,2026-04-19 7:15,27.4,81.3,Patchy light rain with thunder,...,32.05,35.15,2,3,6:14 AM,6:27 PM,7:47 AM,8:32 PM,Waxing Crescent,3
136296,Equatorial Guinea,Malabo,3.7500,8.7833,Africa/Malabo,1776579300,2026-04-19 7:15,27.3,81.1,Patchy light rain with thunder,...,15.85,19.45,2,2,6:17 AM,6:30 PM,7:51 AM,8:35 PM,Waxing Crescent,3
136469,Cameroon,Douala,4.0503,9.7000,Africa/Douala,1776666600,2026-04-20 7:30,24.1,75.4,Patchy light rain with thunder,...,27.85,31.95,2,3,6:13 AM,6:27 PM,8:48 AM,9:37 PM,Waxing Crescent,9


### 4

In [367]:
df[df['moon_illumination']==0]

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
4285,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1717682400,2024-06-06 18:30,18.0,64.3,Patchy rain nearby,...,3.90,8.20,1,1,4:40 AM,7:04 PM,4:02 AM,7:18 PM,New Moon,0
4286,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1717682400,2024-06-06 16:00,32.0,89.6,Sunny,...,4.50,5.10,1,1,5:08 AM,8:11 PM,4:31 AM,8:41 PM,New Moon,0
4287,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1717682400,2024-06-06 15:00,32.0,89.6,Sunny,...,16.40,34.80,2,2,5:29 AM,8:04 PM,4:58 AM,8:33 PM,New Moon,0
4288,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1717682400,2024-06-06 16:00,22.5,72.6,Patchy rain nearby,...,2.20,3.40,1,1,6:17 AM,9:29 PM,5:43 AM,10:03 PM,New Moon,0
4289,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1717682400,2024-06-06 15:00,26.0,78.8,Partly cloudy,...,7.30,25.20,1,1,6:17 AM,5:55 PM,6:04 AM,5:55 PM,New Moon,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136237,Vanuatu,Port Vila,-17.7333,168.3167,Pacific/Efate,1776493800,2026-04-18 17:30,26.0,78.8,Light rain shower,...,5.55,8.45,1,1,5:57 AM,5:36 PM,6:18 AM,5:57 PM,New Moon,0
136239,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776493800,2026-04-18 13:30,27.3,81.1,Partly cloudy,...,106.35,109.55,4,10,5:36 AM,6:16 PM,5:48 AM,7:16 PM,Waxing Crescent,0
136240,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776493800,2026-04-18 9:30,21.4,70.5,Sunny,...,22.05,114.75,2,2,5:47 AM,6:18 PM,6:10 AM,7:24 PM,Waxing Crescent,0
136241,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776493800,2026-04-18 8:30,20.6,69.0,Partly Cloudy,...,5.65,5.75,1,1,6:15 AM,5:58 PM,7:00 AM,6:40 PM,Waxing Crescent,0


### 5

In [368]:
df[(df['temperature_celsius']==20) & (df['condition_text']=='Sunny')]

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
119,Namibia,Windhoek,-22.5700,17.0800,Africa/Windhoek,1715849100,2024-05-16 10:45,20.0,68.0,Sunny,...,2.10,5.50,1,1,7:18 AM,6:18 PM,2:14 PM,1:06 AM,Waxing Gibbous,55
161,South Africa,Pretoria,-25.7500,28.1900,Africa/Johannesburg,1715849100,2024-05-16 10:45,20.0,68.0,Sunny,...,78.10,108.30,4,10,6:38 AM,5:29 PM,1:32 PM,12:16 AM,Waxing Gibbous,55
168,Sweden,Stockholm,59.3300,18.0500,Europe/Stockholm,1715849100,2024-05-16 10:45,20.0,68.0,Sunny,...,4.20,6.10,1,1,4:12 AM,9:18 PM,12:22 PM,3:11 AM,Waxing Gibbous,55
250,Estonia,Tallinn,59.4300,24.7300,Europe/Tallinn,1715868000,2024-05-16 17:00,20.0,68.0,Sunny,...,8.00,8.40,1,1,4:44 AM,9:52 PM,12:54 PM,3:45 AM,Waxing Gibbous,55
337,Romania,Bucharest,44.4300,26.1000,Europe/Bucharest,1715868900,2024-05-16 17:15,20.0,68.0,Sunny,...,2.90,3.70,1,1,5:47 AM,8:38 PM,1:27 PM,2:54 AM,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
135359,Kyrghyzstan,Bishkek,42.8731,74.6003,Asia/Bishkek,1776150000,2026-04-14 13:00,20.0,68.0,Sunny,...,9.75,12.55,1,1,6:22 AM,7:43 PM,4:44 AM,4:09 PM,Waning Crescent,17
135376,Mauritania,Nouakchott,18.0864,-15.9753,Africa/Nouakchott,1776150000,2026-04-14 7:00,20.0,68.0,Sunny,...,45.25,236.95,3,5,6:48 AM,7:21 PM,4:33 AM,4:39 PM,Waning Crescent,15
135543,Iraq,Baghdad,33.3386,44.3939,Asia/Baghdad,1776234600,2026-04-15 9:30,20.0,68.1,Sunny,...,8.85,13.35,1,1,5:33 AM,6:33 PM,4:05 AM,4:26 PM,Waning Crescent,9
136046,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776407400,2026-04-17 8:30,20.0,67.9,Sunny,...,5.05,5.35,1,1,6:14 AM,5:58 PM,5:58 AM,5:52 PM,New Moon,1


### Selecting Rows
- df.loc()
- df.iloc()

In [369]:
df.iloc[4]
# single row

country                                  Angola
location_name                            Luanda
latitude                                  -8.84
longitude                                 13.23
timezone                          Africa/Luanda
last_updated_epoch                   1715849100
last_updated                    2024-05-16 9:45
temperature_celsius                        26.0
temperature_fahrenheit                     78.8
condition_text                    Partly cloudy
wind_mph                                    8.1
wind_kph                                   13.0
wind_degree                                 150
wind_direction                              SSE
pressure_mb                                1011
pressure_in                               29.85
precip_mm                                   0.0
precip_in                                   0.0
humidity                                     89
cloud                                        50
feels_like_celsius                      

In [370]:
df.iloc[1:4]
# portion

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
1,Albania,Tirana,41.33,19.82,Europe/Tirane,1715849100,2024-05-16 10:45,19.0,66.2,Partly cloudy,...,1.1,2.0,1,1,5:21 AM,7:54 PM,12:58 PM,2:14 AM,Waxing Gibbous,55
2,Algeria,Algiers,36.76,3.05,Africa/Algiers,1715849100,2024-05-16 9:45,23.0,73.4,Sunny,...,10.4,18.4,1,1,5:40 AM,7:50 PM,1:15 PM,2:14 AM,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.50,1.52,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.7,0.9,1,1,6:31 AM,9:11 PM,2:12 PM,3:31 AM,Waxing Gibbous,55


#### loc method

In [371]:
df.loc[100,'sunrise']

'5:49 AM'

### Rename columns
- df.rename(columns={'old_name':'new_name},inplace=True)

In [372]:
df.rename(columns={'location_name':'location', 'condition_text':'condition'},inplace=True)

In [373]:
df

,country,location,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1715849100,2024-05-16 13:15,26.6,79.8,Partly Cloudy,...,8.40,26.60,1,1,4:50 AM,6:50 PM,12:12 PM,1:11 AM,Waxing Gibbous,55
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1715849100,2024-05-16 10:45,19.0,66.2,Partly cloudy,...,1.10,2.00,1,1,5:21 AM,7:54 PM,12:58 PM,2:14 AM,Waxing Gibbous,55
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 9:45,23.0,73.4,Sunny,...,10.40,18.40,1,1,5:40 AM,7:50 PM,1:15 PM,2:14 AM,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.70,0.90,1,1,6:31 AM,9:11 PM,2:12 PM,3:31 AM,Waxing Gibbous,55
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1715849100,2024-05-16 9:45,26.0,78.8,Partly cloudy,...,183.40,262.30,5,10,6:12 AM,5:55 PM,1:17 PM,12:38 AM,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 2:30,17.5,63.6,Partly Cloudy,...,7.85,9.35,1,1,6:14 AM,6:39 PM,8:54 AM,10:12 PM,Waxing Crescent,10
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30,28.3,82.9,Cloudy,...,36.25,39.45,2,4,5:34 AM,6:17 PM,7:30 AM,9:34 PM,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 9:30,21.0,69.7,Sunny,...,12.85,51.65,1,2,5:46 AM,6:18 PM,8:00 AM,9:39 PM,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 8:30,21.5,70.8,Sunny,...,5.55,5.85,1,1,6:15 AM,5:56 PM,9:13 AM,8:35 PM,Waxing Crescent,9


In [374]:
df.rename(columns={'air_quality_us-epa-index':'epa_aqi', 'air_quality_gb-defra-index':'daqi'},inplace=True)

In [375]:
df

,country,location,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition,...,air_quality_PM2.5,air_quality_PM10,epa_aqi,daqi,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1715849100,2024-05-16 13:15,26.6,79.8,Partly Cloudy,...,8.40,26.60,1,1,4:50 AM,6:50 PM,12:12 PM,1:11 AM,Waxing Gibbous,55
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1715849100,2024-05-16 10:45,19.0,66.2,Partly cloudy,...,1.10,2.00,1,1,5:21 AM,7:54 PM,12:58 PM,2:14 AM,Waxing Gibbous,55
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 9:45,23.0,73.4,Sunny,...,10.40,18.40,1,1,5:40 AM,7:50 PM,1:15 PM,2:14 AM,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.70,0.90,1,1,6:31 AM,9:11 PM,2:12 PM,3:31 AM,Waxing Gibbous,55
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1715849100,2024-05-16 9:45,26.0,78.8,Partly cloudy,...,183.40,262.30,5,10,6:12 AM,5:55 PM,1:17 PM,12:38 AM,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 2:30,17.5,63.6,Partly Cloudy,...,7.85,9.35,1,1,6:14 AM,6:39 PM,8:54 AM,10:12 PM,Waxing Crescent,10
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30,28.3,82.9,Cloudy,...,36.25,39.45,2,4,5:34 AM,6:17 PM,7:30 AM,9:34 PM,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 9:30,21.0,69.7,Sunny,...,12.85,51.65,1,2,5:46 AM,6:18 PM,8:00 AM,9:39 PM,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 8:30,21.5,70.8,Sunny,...,5.55,5.85,1,1,6:15 AM,5:56 PM,9:13 AM,8:35 PM,Waxing Crescent,9


In [376]:
df['moon_phase'].value_counts()

moon_phase
Waning Crescent    29886
Waxing Crescent    29481
Waxing Gibbous     29382
Waning Gibbous     28477
Full Moon           4933
Last Quarter        4920
New Moon            4876
First Quarter       4678
Name: count, dtype: int64

In [377]:
df[df['temperature_celsius']<=-29]

,country,location,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition,...,air_quality_PM2.5,air_quality_PM10,epa_aqi,daqi,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
118990,Mongolia,Ulaanbaatar,47.9167,106.9167,Asia/Ulaanbaatar,1768719600,2026-01-18 15:00,-29.8,-21.6,Mist,...,9.15,9.35,1,1,8:35 AM,5:31 PM,8:39 AM,4:36 PM,Waning Crescent,1


# Data Cleaning

### cleaning Numerical columns with characters

In [378]:
df.head(1)

,country,location,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition,...,air_quality_PM2.5,air_quality_PM10,epa_aqi,daqi,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.52,69.18,Asia/Kabul,1715849100,2024-05-16 13:15,26.6,79.8,Partly Cloudy,...,8.4,26.6,1,1,4:50 AM,6:50 PM,12:12 PM,1:11 AM,Waxing Gibbous,55


In [379]:
df.dtypes

country                          object
location                         object
latitude                        float64
longitude                       float64
timezone                         object
last_updated_epoch                int64
last_updated                     object
temperature_celsius             float64
temperature_fahrenheit          float64
condition                        object
wind_mph                        float64
wind_kph                        float64
wind_degree                       int64
wind_direction                   object
pressure_mb                       int64
pressure_in                     float64
precip_mm                       float64
precip_in                       float64
humidity                          int64
cloud                             int64
feels_like_celsius              float64
feels_like_fahrenheit           float64
visibility_km                   float64
visibility_miles                  int64
uv_index                        float64


In [380]:
df['sunrise']=pd.to_datetime(
    df['sunrise']
    .astype(str)
    .str.replace(r'[^0-9.-]','',regex=True)
    .replace('',None),
    errors='coerce')

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_16652\667980494.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df['sunrise']=pd.to_datetime(


In [381]:
df.head(10)

,country,location,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition,...,air_quality_PM2.5,air_quality_PM10,epa_aqi,daqi,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.52,69.18,Asia/Kabul,1715849100,2024-05-16 13:15,26.6,79.8,Partly Cloudy,...,8.4,26.6,1,1,NaT,6:50 PM,12:12 PM,1:11 AM,Waxing Gibbous,55
1,Albania,Tirana,41.33,19.82,Europe/Tirane,1715849100,2024-05-16 10:45,19.0,66.2,Partly cloudy,...,1.1,2.0,1,1,NaT,7:54 PM,12:58 PM,2:14 AM,Waxing Gibbous,55
2,Algeria,Algiers,36.76,3.05,Africa/Algiers,1715849100,2024-05-16 9:45,23.0,73.4,Sunny,...,10.4,18.4,1,1,NaT,7:50 PM,1:15 PM,2:14 AM,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.50,1.52,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.7,0.9,1,1,NaT,9:11 PM,2:12 PM,3:31 AM,Waxing Gibbous,55
4,Angola,Luanda,-8.84,13.23,Africa/Luanda,1715849100,2024-05-16 9:45,26.0,78.8,Partly cloudy,...,183.4,262.3,5,10,NaT,5:55 PM,1:17 PM,12:38 AM,Waxing Gibbous,55
5,Antigua and Barbuda,Saint John's,17.12,-61.85,America/Antigua,1715849100,2024-05-16 4:45,26.0,78.8,Partly cloudy,...,1.2,4.5,1,1,NaT,6:32 PM,1:05 PM,1:14 AM,Waxing Gibbous,55
6,Argentina,Buenos Aires,-34.59,-58.67,America/Argentina/Buenos_Aires,1715849100,2024-05-16 5:45,8.0,46.4,Clear,...,4.0,5.3,1,1,NaT,5:59 PM,2:36 PM,1:04 AM,Waxing Gibbous,55
7,Armenia,Yerevan,40.18,44.51,Asia/Yerevan,1715849100,2024-05-16 12:45,19.0,66.2,Partly cloudy,...,0.8,0.9,1,1,NaT,8:12 PM,1:17 PM,2:31 AM,Waxing Gibbous,55
8,Australia,Canberra,-35.28,149.22,Australia/Sydney,1715849100,2024-05-16 18:45,9.0,48.2,Clear,...,3.7,5.4,1,1,NaT,5:07 PM,1:31 PM,No moonset,Waxing Gibbous,55
9,Austria,Vienna,48.20,16.37,Europe/Vienna,1715849100,2024-05-16 10:45,16.0,60.8,Partly cloudy,...,3.7,4.4,1,1,NaT,8:29 PM,1:00 PM,2:42 AM,Waxing Gibbous,55


In [382]:
df.dtypes

country                                 object
location                                object
latitude                               float64
longitude                              float64
timezone                                object
last_updated_epoch                       int64
last_updated                            object
temperature_celsius                    float64
temperature_fahrenheit                 float64
condition                               object
wind_mph                               float64
wind_kph                               float64
wind_degree                              int64
wind_direction                          object
pressure_mb                              int64
pressure_in                            float64
precip_mm                              float64
precip_in                              float64
humidity                                 int64
cloud                                    int64
feels_like_celsius                     float64
feels_like_fa

- Changing last_updated object/string datatype to datetime datatype

In [383]:
df['last_updated']= pd.to_datetime(df['last_updated'],
                                   errors='coerce')

In [384]:
df.head(5)

,country,location,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition,...,air_quality_PM2.5,air_quality_PM10,epa_aqi,daqi,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.52,69.18,Asia/Kabul,1715849100,2024-05-16 13:15:00,26.6,79.8,Partly Cloudy,...,8.4,26.6,1,1,NaT,6:50 PM,12:12 PM,1:11 AM,Waxing Gibbous,55
1,Albania,Tirana,41.33,19.82,Europe/Tirane,1715849100,2024-05-16 10:45:00,19.0,66.2,Partly cloudy,...,1.1,2.0,1,1,NaT,7:54 PM,12:58 PM,2:14 AM,Waxing Gibbous,55
2,Algeria,Algiers,36.76,3.05,Africa/Algiers,1715849100,2024-05-16 09:45:00,23.0,73.4,Sunny,...,10.4,18.4,1,1,NaT,7:50 PM,1:15 PM,2:14 AM,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.50,1.52,Europe/Andorra,1715849100,2024-05-16 10:45:00,6.3,43.3,Light drizzle,...,0.7,0.9,1,1,NaT,9:11 PM,2:12 PM,3:31 AM,Waxing Gibbous,55
4,Angola,Luanda,-8.84,13.23,Africa/Luanda,1715849100,2024-05-16 09:45:00,26.0,78.8,Partly cloudy,...,183.4,262.3,5,10,NaT,5:55 PM,1:17 PM,12:38 AM,Waxing Gibbous,55


In [385]:
df.dtypes

country                                 object
location                                object
latitude                               float64
longitude                              float64
timezone                                object
last_updated_epoch                       int64
last_updated                    datetime64[ns]
temperature_celsius                    float64
temperature_fahrenheit                 float64
condition                               object
wind_mph                               float64
wind_kph                               float64
wind_degree                              int64
wind_direction                          object
pressure_mb                              int64
pressure_in                            float64
precip_mm                              float64
precip_in                              float64
humidity                                 int64
cloud                                    int64
feels_like_celsius                     float64
feels_like_fa

In [386]:
df.tail(1)

,country,location,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition,...,air_quality_PM2.5,air_quality_PM10,epa_aqi,daqi,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
136632,Zimbabwe,Harare,-17.8178,31.0447,Africa/Harare,1776666600,2026-04-20 08:30:00,20.0,68.1,Sunny,...,7.85,8.05,1,1,NaT,5:43 PM,9:07 AM,8:18 PM,Waxing Crescent,9


In [387]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 136633 entries, 0 to 136632
Data columns (total 41 columns):
 #   Column                        Non-Null Count   Dtype         
---  ------                        --------------   -----         
 0   country                       136633 non-null  object        
 1   location                      136633 non-null  object        
 2   latitude                      136633 non-null  float64       
 3   longitude                     136633 non-null  float64       
 4   timezone                      136633 non-null  object        
 5   last_updated_epoch            136633 non-null  int64         
 6   last_updated                  136633 non-null  datetime64[ns]
 7   temperature_celsius           136633 non-null  float64       
 8   temperature_fahrenheit        136633 non-null  float64       
 9   condition                     136633 non-null  object        
 10  wind_mph                      136633 non-null  float64       
 11  wind_kph     

In [388]:
df.head(1
)

,country,location,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition,...,air_quality_PM2.5,air_quality_PM10,epa_aqi,daqi,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.52,69.18,Asia/Kabul,1715849100,2024-05-16 13:15:00,26.6,79.8,Partly Cloudy,...,8.4,26.6,1,1,NaT,6:50 PM,12:12 PM,1:11 AM,Waxing Gibbous,55


In [389]:
df[['temperature_fahrenheit', 'temperature_celsius']]

,temperature_fahrenheit,temperature_celsius
0,79.8,26.6
1,66.2,19.0
2,73.4,23.0
3,43.3,6.3
4,78.8,26.0
...,...,...
136628,63.6,17.5
136629,82.9,28.3
136630,69.7,21.0
136631,70.8,21.5


### Deleting unwanted columns

In [390]:
df.drop(columns=['sunrise','sunset','moonrise','moonset','moon_phase','moon_illumination'],inplace=True)

In [391]:
df

,country,location,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition,...,gust_mph,gust_kph,air_quality_Carbon_Monoxide,air_quality_Ozone,air_quality_Nitrogen_dioxide,air_quality_Sulphur_dioxide,air_quality_PM2.5,air_quality_PM10,epa_aqi,daqi
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1715849100,2024-05-16 13:15:00,26.6,79.8,Partly Cloudy,...,9.5,15.3,277.00,103.0,1.10,0.20,8.40,26.60,1,1
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1715849100,2024-05-16 10:45:00,19.0,66.2,Partly cloudy,...,11.4,18.4,193.60,97.3,0.90,0.10,1.10,2.00,1,1
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 09:45:00,23.0,73.4,Sunny,...,13.9,22.3,540.70,12.2,65.10,13.40,10.40,18.40,1,1
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1715849100,2024-05-16 10:45:00,6.3,43.3,Light drizzle,...,8.5,13.7,170.20,64.4,1.60,0.20,0.70,0.90,1,1
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1715849100,2024-05-16 09:45:00,26.0,78.8,Partly cloudy,...,12.5,20.2,2964.00,19.0,72.70,31.50,183.40,262.30,5,10
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 02:30:00,17.5,63.6,Partly Cloudy,...,6.7,10.8,577.85,35.0,8.15,1.65,7.85,9.35,1,1
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30:00,28.3,82.9,Cloudy,...,12.7,20.5,351.85,35.0,34.45,21.05,36.25,39.45,2,4
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 09:30:00,21.0,69.7,Sunny,...,5.7,9.1,149.85,58.0,5.85,2.05,12.85,51.65,1,2
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 08:30:00,21.5,70.8,Sunny,...,11.1,17.8,160.85,40.0,2.85,2.05,5.55,5.85,1,1


In [392]:
df.drop(columns=['temperature_fahrenheit','feels_like_fahrenheit','wind_mph','precip_in','visibility_miles','gust_mph','pressure_in'],inplace=True)

In [393]:
df

,country,location,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,condition,wind_kph,...,uv_index,gust_kph,air_quality_Carbon_Monoxide,air_quality_Ozone,air_quality_Nitrogen_dioxide,air_quality_Sulphur_dioxide,air_quality_PM2.5,air_quality_PM10,epa_aqi,daqi
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1715849100,2024-05-16 13:15:00,26.6,Partly Cloudy,13.3,...,7.0,15.3,277.00,103.0,1.10,0.20,8.40,26.60,1,1
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1715849100,2024-05-16 10:45:00,19.0,Partly cloudy,11.2,...,5.0,18.4,193.60,97.3,0.90,0.10,1.10,2.00,1,1
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 09:45:00,23.0,Sunny,15.1,...,5.0,22.3,540.70,12.2,65.10,13.40,10.40,18.40,1,1
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1715849100,2024-05-16 10:45:00,6.3,Light drizzle,11.9,...,2.0,13.7,170.20,64.4,1.60,0.20,0.70,0.90,1,1
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1715849100,2024-05-16 09:45:00,26.0,Partly cloudy,13.0,...,8.0,20.2,2964.00,19.0,72.70,31.50,183.40,262.30,5,10
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 02:30:00,17.5,Partly Cloudy,3.6,...,0.0,10.8,577.85,35.0,8.15,1.65,7.85,9.35,1,1
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30:00,28.3,Cloudy,17.6,...,5.5,20.5,351.85,35.0,34.45,21.05,36.25,39.45,2,4
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 09:30:00,21.0,Sunny,7.9,...,6.5,9.1,149.85,58.0,5.85,2.05,12.85,51.65,1,2
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 08:30:00,21.5,Sunny,13.0,...,1.6,17.8,160.85,40.0,2.85,2.05,5.55,5.85,1,1


In [394]:
df.drop(columns=['last_updated_epoch'],inplace=True)

In [395]:
df

,country,location,latitude,longitude,timezone,last_updated,temperature_celsius,condition,wind_kph,wind_degree,...,uv_index,gust_kph,air_quality_Carbon_Monoxide,air_quality_Ozone,air_quality_Nitrogen_dioxide,air_quality_Sulphur_dioxide,air_quality_PM2.5,air_quality_PM10,epa_aqi,daqi
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,2024-05-16 13:15:00,26.6,Partly Cloudy,13.3,338,...,7.0,15.3,277.00,103.0,1.10,0.20,8.40,26.60,1,1
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,2024-05-16 10:45:00,19.0,Partly cloudy,11.2,320,...,5.0,18.4,193.60,97.3,0.90,0.10,1.10,2.00,1,1
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,2024-05-16 09:45:00,23.0,Sunny,15.1,280,...,5.0,22.3,540.70,12.2,65.10,13.40,10.40,18.40,1,1
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,2024-05-16 10:45:00,6.3,Light drizzle,11.9,215,...,2.0,13.7,170.20,64.4,1.60,0.20,0.70,0.90,1,1
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,2024-05-16 09:45:00,26.0,Partly cloudy,13.0,150,...,8.0,20.2,2964.00,19.0,72.70,31.50,183.40,262.30,5,10
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,2026-04-20 02:30:00,17.5,Partly Cloudy,3.6,122,...,0.0,10.8,577.85,35.0,8.15,1.65,7.85,9.35,1,1
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,2026-04-20 13:30:00,28.3,Cloudy,17.6,151,...,5.5,20.5,351.85,35.0,34.45,21.05,36.25,39.45,2,4
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,2026-04-20 09:30:00,21.0,Sunny,7.9,135,...,6.5,9.1,149.85,58.0,5.85,2.05,12.85,51.65,1,2
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,2026-04-20 08:30:00,21.5,Sunny,13.0,58,...,1.6,17.8,160.85,40.0,2.85,2.05,5.55,5.85,1,1


In [396]:
df.shape

(136633, 27)

### 27 columns from 41

In [397]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 136633 entries, 0 to 136632
Data columns (total 27 columns):
 #   Column                        Non-Null Count   Dtype         
---  ------                        --------------   -----         
 0   country                       136633 non-null  object        
 1   location                      136633 non-null  object        
 2   latitude                      136633 non-null  float64       
 3   longitude                     136633 non-null  float64       
 4   timezone                      136633 non-null  object        
 5   last_updated                  136633 non-null  datetime64[ns]
 6   temperature_celsius           136633 non-null  float64       
 7   condition                     136633 non-null  object        
 8   wind_kph                      136633 non-null  float64       
 9   wind_degree                   136633 non-null  int64         
 10  wind_direction                136633 non-null  object        
 11  pressure_mb  

### Handling missing values

In [398]:
df.isnull().sum()

country                         0
location                        0
latitude                        0
longitude                       0
timezone                        0
last_updated                    0
temperature_celsius             0
condition                       0
wind_kph                        0
wind_degree                     0
wind_direction                  0
pressure_mb                     0
precip_mm                       0
humidity                        0
cloud                           0
feels_like_celsius              0
visibility_km                   0
uv_index                        0
gust_kph                        0
air_quality_Carbon_Monoxide     0
air_quality_Ozone               0
air_quality_Nitrogen_dioxide    0
air_quality_Sulphur_dioxide     0
air_quality_PM2.5               0
air_quality_PM10                0
epa_aqi                         0
daqi                            0
dtype: int64

- finding median and mode to fill in missing data
- no missing data in this dataset

In [399]:
df['temperature_celsius'].median()

23.9

In [400]:
df['timezone'].mode()

0    Asia/Bangkok
Name: timezone, dtype: object

Finding mean, median, and mode of each columns

In [401]:
print(df.select_dtypes(include='number').mean())
# To find mean of numerical columns

latitude                          19.211173
longitude                         21.956700
temperature_celsius               21.268191
wind_kph                          12.897045
wind_degree                      168.797348
pressure_mb                     1014.064304
precip_mm                          0.134168
humidity                          66.661553
cloud                             39.976228
feels_like_celsius                22.120190
visibility_km                      9.507450
uv_index                           3.313816
gust_kph                          18.242712
air_quality_Carbon_Monoxide      459.169559
air_quality_Ozone                 57.997988
air_quality_Nitrogen_dioxide      14.974117
air_quality_Sulphur_dioxide       10.315061
air_quality_PM2.5                 24.149276
air_quality_PM10                  48.293201
epa_aqi                            1.697430
daqi                               2.601275
dtype: float64


In [402]:
print(df.select_dtypes(include='number').median())
# To find median of numerical columns


latitude                          17.2500
longitude                         23.2361
temperature_celsius               23.9000
wind_kph                          10.8000
wind_degree                      161.0000
pressure_mb                     1014.0000
precip_mm                          0.0000
humidity                          72.0000
cloud                             30.0000
feels_like_celsius                25.1000
visibility_km                     10.0000
uv_index                           1.9000
gust_kph                          15.3000
air_quality_Carbon_Monoxide      293.7000
air_quality_Ozone                 55.0000
air_quality_Nitrogen_dioxide       5.6500
air_quality_Sulphur_dioxide        2.4050
air_quality_PM2.5                 14.1500
air_quality_PM10                  19.9800
epa_aqi                            1.0000
daqi                               2.0000
dtype: float64


In [403]:
c=df.select_dtypes(include='object')
for i in c:
    print(i,df[i].mode())
    print()
# To find mode of categorical columns
# There is many locations because it

country 0    Bulgaria
Name: country, dtype: object

location 0            Accra
1            Amman
2           Asmara
3          Baghdad
4             Bern
5        Bujumbura
6            Dakar
7            Kabul
8             Kyiv
9           Malabo
10       N'djamena
11             Nan
12           Sanaa
13            Suva
14        Tashkent
15           Tokyo
16        Valletta
17    Vatican City
18          Warsaw
Name: location, dtype: object

timezone 0    Asia/Bangkok
Name: timezone, dtype: object

condition 0    Partly cloudy
Name: condition, dtype: object

wind_direction 0    E
Name: wind_direction, dtype: object



### Handling duplicates

In [404]:
df.duplicated().sum()

np.int64(1)

In [405]:
print('before',df.shape)
df=df.drop_duplicates()
print('after',df.shape)

before (136633, 27)
after (136632, 27)


In [406]:
df.duplicated().sum()

np.int64(0)

In [407]:
df.head(1)

,country,location,latitude,longitude,timezone,last_updated,temperature_celsius,condition,wind_kph,wind_degree,...,uv_index,gust_kph,air_quality_Carbon_Monoxide,air_quality_Ozone,air_quality_Nitrogen_dioxide,air_quality_Sulphur_dioxide,air_quality_PM2.5,air_quality_PM10,epa_aqi,daqi
0,Afghanistan,Kabul,34.52,69.18,Asia/Kabul,2024-05-16 13:15:00,26.6,Partly Cloudy,13.3,338,...,7.0,15.3,277.0,103.0,1.1,0.2,8.4,26.6,1,1


In [408]:
df.shape

(136632, 27)

In [409]:
df.columns

Index(['country', 'location', 'latitude', 'longitude', 'timezone',
       'last_updated', 'temperature_celsius', 'condition', 'wind_kph',
       'wind_degree', 'wind_direction', 'pressure_mb', 'precip_mm', 'humidity',
       'cloud', 'feels_like_celsius', 'visibility_km', 'uv_index', 'gust_kph',
       'air_quality_Carbon_Monoxide', 'air_quality_Ozone',
       'air_quality_Nitrogen_dioxide', 'air_quality_Sulphur_dioxide',
       'air_quality_PM2.5', 'air_quality_PM10', 'epa_aqi', 'daqi'],
      dtype='object')

In [410]:
df.head(1)

,country,location,latitude,longitude,timezone,last_updated,temperature_celsius,condition,wind_kph,wind_degree,...,uv_index,gust_kph,air_quality_Carbon_Monoxide,air_quality_Ozone,air_quality_Nitrogen_dioxide,air_quality_Sulphur_dioxide,air_quality_PM2.5,air_quality_PM10,epa_aqi,daqi
0,Afghanistan,Kabul,34.52,69.18,Asia/Kabul,2024-05-16 13:15:00,26.6,Partly Cloudy,13.3,338,...,7.0,15.3,277.0,103.0,1.1,0.2,8.4,26.6,1,1


#### Renaming columns

- temperature_celsius → temp_c
- air_quality_PM2.5 → pm_2.5
- air_quality_PM10 → pm_10
- air_quality_Carbon_Monoxide → co
- air_quality_Ozone → o3
- air_quality_Nitrogen_dioxide → no2
- air_quality_Sulphur_dioxide → so2


In [411]:
df.rename(columns={
    'air_quality_Carbon_Monoxide': 'co',
    'air_quality_Ozone': 'o3',
    'air_quality_Nitrogen_dioxide': 'no2',
    'air_quality_Sulphur_dioxide': 'so2' ,
    'air_quality_PM2.5': 'pm_2.5' ,
    'air_quality_PM10' : 'pm_10',
    'temperature_celsius' : 'temp_c'
}, inplace=True)

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_16652\54472390.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.rename(columns={


#### Saving cleaned csv file

In [412]:
df.to_csv('weather_cleaned.csv', index=False)

In [413]:
df_new = pd.read_csv('weather_cleaned.csv')

In [414]:
df_new.head(1)

,country,location,latitude,longitude,timezone,last_updated,temp_c,condition,wind_kph,wind_degree,...,uv_index,gust_kph,co,o3,no2,so2,pm_2.5,pm_10,epa_aqi,daqi
0,Afghanistan,Kabul,34.52,69.18,Asia/Kabul,2024-05-16 13:15:00,26.6,Partly Cloudy,13.3,338,...,7.0,15.3,277.0,103.0,1.1,0.2,8.4,26.6,1,1


In [415]:
df_new.dtypes

country                object
location               object
latitude              float64
longitude             float64
timezone               object
last_updated           object
temp_c                float64
condition              object
wind_kph              float64
wind_degree             int64
wind_direction         object
pressure_mb             int64
precip_mm             float64
humidity                int64
cloud                   int64
feels_like_celsius    float64
visibility_km         float64
uv_index              float64
gust_kph              float64
co                    float64
o3                    float64
no2                   float64
so2                   float64
pm_2.5                float64
pm_10                 float64
epa_aqi                 int64
daqi                    int64
dtype: object

# Summary Statistics 

- mean
- median
- mode
- standard deviation

### Mean(Average)

In [416]:
print(df.select_dtypes(include='number').mean())
# To find mean of numerical columns

latitude                19.211176
longitude               21.956123
temp_c                  21.268111
wind_kph                12.897108
wind_degree            168.797280
pressure_mb           1014.064297
precip_mm                0.134169
humidity                66.661726
cloud                   39.976338
feels_like_celsius      22.120121
visibility_km            9.507446
uv_index                 3.313791
gust_kph                18.242809
co                     459.169180
o3                      57.998222
no2                     14.974119
so2                     10.315115
pm_2.5                  24.149295
pm_10                   48.293397
epa_aqi                  1.697428
daqi                     2.601279
dtype: float64


### Median

In [417]:
print(df.select_dtypes(include='number').median())
# To find median of numerical columns


latitude                17.2500
longitude               23.2361
temp_c                  23.9000
wind_kph                10.8000
wind_degree            161.0000
pressure_mb           1014.0000
precip_mm                0.0000
humidity                72.0000
cloud                   30.0000
feels_like_celsius      25.1000
visibility_km           10.0000
uv_index                 1.9000
gust_kph                15.3000
co                     293.7000
o3                      55.0000
no2                      5.6500
so2                      2.4050
pm_2.5                  14.1500
pm_10                   19.9800
epa_aqi                  1.0000
daqi                     2.0000
dtype: float64


### Mode

In [418]:
c=df.select_dtypes(include='object')
for i in c:
    print(i,df[i].mode())
    print()
# To find mode of categorical columns
# There is many locations because it

country 0    Bulgaria
Name: country, dtype: object

location 0            Accra
1            Amman
2           Asmara
3          Baghdad
4             Bern
5        Bujumbura
6            Dakar
7            Kabul
8             Kyiv
9           Malabo
10       N'djamena
11           Sanaa
12            Suva
13        Tashkent
14           Tokyo
15        Valletta
16    Vatican City
17          Warsaw
Name: location, dtype: object

timezone 0    Asia/Bangkok
Name: timezone, dtype: object

condition 0    Partly cloudy
Name: condition, dtype: object

wind_direction 0    E
Name: wind_direction, dtype: object



### Standard Deviation

In [419]:
print(df.select_dtypes(include='number').std())

latitude               24.416057
longitude              65.787067
temp_c                  9.687157
wind_kph               11.587170
wind_degree           103.752702
pressure_mb            10.327589
precip_mm               0.562252
humidity               23.871459
cloud                  34.096311
feels_like_celsius     11.590592
visibility_km           2.691788
uv_index                3.548004
gust_kph               13.793096
co                    755.709390
o3                     30.711697
no2                    23.649746
so2                    35.505696
pm_2.5                 36.761319
pm_10                 148.886879
epa_aqi                 0.939249
daqi                    2.440300
dtype: float64


### Table to compare mean, median, & Std

In [420]:
stats_df = df.select_dtypes(include='number').agg(['mean', 'median', 'std']).T
stats_df

,mean,median,std
latitude,19.211176,17.2500,24.416057
longitude,21.956123,23.2361,65.787067
temp_c,21.268111,23.9000,9.687157
wind_kph,12.897108,10.8000,11.587170
wind_degree,168.797280,161.0000,103.752702
pressure_mb,1014.064297,1014.0000,10.327589
precip_mm,0.134169,0.0000,0.562252
humidity,66.661726,72.0000,23.871459
cloud,39.976338,30.0000,34.096311
feels_like_celsius,22.120121,25.1000,11.590592


In [426]:
df.describe().T

,count,mean,min,25%,50%,75%,max,std
latitude,136632.0,19.211176,-41.3,4.0503,17.25,40.4,64.15,24.416057
longitude,136632.0,21.956123,-175.2,-6.8361,23.2361,49.8822,179.22,65.787067
last_updated,136632,2025-05-03 09:03:41.923854080,2024-05-16 01:45:00,2024-11-09 05:00:00,2025-05-03 11:00:00,2025-10-25 14:22:30,2026-04-20 19:30:00,NaN
temp_c,136632.0,21.268111,-29.8,15.9,23.9,28.0,49.2,9.687157
wind_kph,136632.0,12.897108,3.6,6.1,10.8,17.6,2963.2,11.58717
wind_degree,136632.0,168.79728,1.0,80.0,161.0,256.0,360.0,103.752702
pressure_mb,136632.0,1014.064297,947.0,1010.0,1014.0,1018.0,3006.0,10.327589
precip_mm,136632.0,0.134169,0.0,0.0,0.0,0.03,42.24,0.562252
humidity,136632.0,66.661726,2.0,51.0,72.0,86.0,100.0,23.871459
cloud,136632.0,39.976338,0.0,0.0,30.0,75.0,100.0,34.096311


### INF0

#### mean and median:

##### Same:
- pressure_mb
- epa_aqi
- daqi
- precip_mm
- visibility_km
- TOTAL = 5

##### Almost same
- feels_like_celcius
- temp_c
- latitude
- longitude
- wind_kph
- gust_kph
- o3
- TOTAL = 7

##### Different
- co
- pm_2.5
- pm_10
- wind_degree
- humidity
- cloud
- uv_index
- so2
- no2
- TOTAL = 9

In [427]:
df.describe(include=object)

,country,location,timezone,condition,wind_direction
count,136632,136632,136632,136632,136632
unique,210,257,199,49,16
top,Bulgaria,Kabul,Asia/Bangkok,Partly cloudy,E
freq,1621,703,2516,39376,12995


### INFO
* The dataset is complete with no missing categorical values.
* There is high geographical diversity (200+ countries and locations).
* Weather conditions are repetitive, with “Partly cloudy” being the most common.
* Wind direction shows a pattern, with East being the most frequent.
* Some locations (like Kabul) appear more often, suggesting the dataset may be unevenly distributed.